### ISSUE TITLE
CPU 100% usage while idling on some backends.

### ISSUE THREAD

**Issue by OceanWolf** [matplotlib/matplotlib#4092]
I get 100% CPU usage just opening a single plot like the following, and 200% running from two separate python instances (dual-core):

```python
import matplotlib.pyplot as plt
plt.figure()
plt.plot(range(5))
plt.show()
```
I have tested this on ``GTK`` and ``GTK3``, both exhibit this, and on ``QT4`` which does not.

**Comment by WeatherGod** [comment]
I have never seen this behavior. We are going to need more information.
What OS, which matplotlib, and versions of GTK. Are the TkAgg and WxAgg
backends suffering from this? How were the packages installed?

**Comment by mdboom** [comment]
I can reproduced this on Gtk and Gtk3 on Linux with matplotlib master, but not 1.4.x.  I'm going to see if `git bisect` can track down the cause of the change.

**Comment by mdboom** [comment]
The first commit where this shows up is 86fd05feafd82d3aad3d6d4f477c6097bfca0fa4, which makes a lot of sense.  @ahaldane: Is there any way to fix the bug that was trying so solve without pegging the CPU?

**Comment by tacaswell** [comment]
Added in PR #3769

Sorry about that....

**Comment by OceanWolf** [comment]
@WeatherGod I use debian8+lxde (which comes with  a CPU usage graph in the dock by default), and yes, I run-master, using gi.__version__ = 3.14.0.  But I guess you don't need that information now ;).

**Comment by WeatherGod** [comment]
My guess is that what is happening is that draw_idle() is constantly being
called, over and over, since the GUI is idled. It wasn't doing that before
because the idle event wasn't being retriggered. One way to confirm this is
to put a print statement in draw_idle and see how often it is called before
and after that commit.

**Comment by OceanWolf** [comment]
Using cprofile (only giving the lines with relatively large numbers, some of course obvious):
19901634 function calls (19901149 primitive calls) in 54.399 seconds

ncalls | tottime | percall | cumtime | percall | filename:lineno(function)
------- | ------- | ------- | ------- | ------- | -------
1 |  33.601 | 33.601 |  54.389 |  54.389 | matplotlib/backends/backend_gtk3.py:73(mainloop)
6622344 |   3.038 |   0.000  |  3.038  |  0.000 | matplotlib/backend_bases.py:1331(__init__)
1 |   0.000 |   0.000 |  54.399 |  54.399 | matplotlib/backend_bases.py:151(__call__)
6622339  | 15.580 |   0.000  | 20.703 |   0.000 | matplotlib/backend_bases.py:1980(idle_event)
1 |   0.000 |   0.000 |  54.399 |  54.399 | <string>:1(<module>)
1 |   0.000 |   0.000 |  54.399 |  54.399 | matplotlib/pyplot.py:137(show)
6622345/6622343  |  2.085 |   0.000 |   2.087 |   0.000 | matplotlib/cbook.py:523(process)

**Comment by WeatherGod** [comment]
So, it looks like there are no callbacks for an Idle event in the gtk
backends... it is just doing no-ops, essentially. I wonder if there is some
way to regulate the frequency of idle events.

**Comment by WeatherGod** [comment]
huh, there is idle_draw and idle_event in backend_gtk3.py. The idle_draw
returns False, making it a one-shot action.

**Comment by OceanWolf** [comment]
Oh, and to generate the above, I called:
```python
import cProfile
# snip
cProfile.run('plt.show()') # Although here I saved data to file and performed stats later.
```
and in case you wonder ``backend_bases.py:1331`` refers to``Event.__init__``, presumably called from the creation of an ``IdleEvent`` class.

So no @WeatherGod, it appears that draw_idle gets called just the once, taking up virtually no time (all other calls give 0.000).  It looks to me that GTK incessantly asks ``Okay, I have done everything you asked, what now?`` to mpl, so basically, it doesn't get any sleep.

I think I have found the problem...
In ``backend_gtk3.py`` we have two methods for idle events going on... the one set up in ``FigureCanvasGTK3.__init__`` and the one in FigureCanvasGTK3.draw_idle``, the former uses the Event paradigm set up with ``FigureCanvasBase`` and as it returns True, creates this loop; the later does the drawing on idle, but does it completely self-contained... I think the solution lies in moulding the two into one.

**Comment by WeatherGod** [comment]
I was referring to the callback ids. there is self._idle_draw_id, and
self._idle_event_id. When draw_idle() is called, it uses
self._idle_draw_id, and is one-shot. While the 'idle_event' (as in the
mpl_connect() name) uses self._idle_event_id, and is not one-shot. The
referenced PR changed that callback from one-shot to recurring (and for
good reason). We need these things separate, as the call to idly draw needs
to be one-shot (I never really delved down into the idle draw code, so I
thought they were one and the same originally).

But yes, the problem is that GTK is constantly saying that it is idle, and
letting us know that. I wonder if there is some sort of rate-limiter that
could be applied?

**Comment by OceanWolf** [comment]
@WeatherGod Sorry, I replied before your last posts showed up and we basically saw the same thing.

However I don't think the problem lies in a rate-limiter... basically as I see it, we have two independent methods of doing the same thing, one that currently works, and one that currently doesn't.

Method 1:
+ Sets up a listener to GTK's idle event in __init__ *bad*
+ Hooking it to CanvasBase's ``idle_event`` method *good* (I think because less GTK specific)
+ Now this generic way used to return None, which cancelled the Event from happening again.
+ But in the commit referenced above, it returned True, which tells GTK to not fire the event any more.  @ahaldane did that to stop an error message from showing when the window tried to remove GTK's idle_event callback.  Because previously it had returned None it meant that that callback had already been removed...

Method 2:
+ Call to ``draw_idle`` in FigureCanvasGTK3 *good place*
+ Handles everything in this GTK specific method *bad* (I think)

So we now see that the "fix" in https://github.com/matplotlib/matplotlib/commit/86fd05feafd82d3aad3d6d4f477c6097bfca0fa4 went after the symptom rather then the cause.

To solve this adequately, one needs to look at how CanvasBase.idle_event should work and make sure it works on GTK (removing the ``return True`` commit).

**Comment by OceanWolf** [comment]
Just read https://github.com/matplotlib/matplotlib/issues/959 and idle_event does nothing at all, (I have just searched through the entire lib dir to see if it gets referenced somewhere, and it does not, even in animation, or in the mpl_toolkits, so...

The solution now looks clear, remove commit https://github.com/matplotlib/matplotlib/commit/a9767f75d61750b4f02793e04725e4f52efda5f2 and remove the lines pertaining to ``idle_event_id`` in ``__init__`` and ``destroy`` of the backend.

@mdboom You said in #959 that you wanted to do something with ``idle_event``, so what do you say to this?

**Comment by ahaldane** [comment]
I agree the problem is `idle_event` gets called too much now. I hadn't realized how often it gets called.

Re the most recent discussion: Maybe I have misunderstood, but don't `draw_idle` and `idle_event` have different purposes?

My limited understanding was that `draw_idle` is a mechanism for matplotlib to say "redraw the screen at the next moment we aren't doing something else". On the other hand `idle_event` gives the user a way to call arbitrary code when the GUI is idle, for example to pre-process data or for "passive animation" (as a comment in the code says).

If that's right, then I think a rate limiter on `idle_event` is a solution, and `draw_idle` should be left alone. Implementing the rate limiter seems a bit ugly - I'm thinking it would involve idle event setting a timer then detaching itself, and the timer reattaches the idle_event listener after waiting.

**Comment by ahaldane** [comment]
@OceanWolf , matplotlib does not use `idle_event` internally, it is meant to be used by the user, for example: http://matplotlib.org/examples/event_handling/idle_and_timeout.html

**Comment by OceanWolf** [comment]
Ahh, okay... but what does it mean now that we have an animation module?

Or... assuming we want to keep this code for now, how about we change the line in __init__ to ``self._idle_event_id = None``, which we can use as a check on whether we have to destroy the event when the canvas gets destroyed, then set up the idle_event in another function that only gets during on the first call to ``CanvasBase.idle_event``.  So as to not alter any of the other backends, you can add a default function to CanvasBase that does nothing which GTK will override.

**Comment by ahaldane** [comment]
I was not aware of the animations module. If it makes `idle_event obsolete`, then maybe we can save ourselves trouble and simply remove the idle event api, solving my old bug plus this one. A github search for "mpl_connect idle_event" turns up nothing, so it doesn't seem to be used much.  The docstring for `CanvasBase.mpl_connect` doesn't even list `idle_event` as a valid event...

As to the idea about only connecting the handler on the first call to `CanvasBase.idle_event`: I don't think I understand. How would `CanvasBase.idle_event` be called in the first place? Currently it is only called in response to a gtk idle event.

**Comment by OceanWolf** [comment]
Ahh sorry, I meant in ``mpl_connect`` so in ``FigureCanvasBase`` you have:
```python
def __init__(...):
    self.using_idle_event = False
    # snip
def mpl_connect(...):
    if s = 'idle_event' and self.using_idle_event is False:
        self.setup_idle_event()
        self.using_idle_event = True
    # snip
def setup_idle_event():
    pass
# snip
```
With this, it pushes the 100% problem on to those who actually use the code.  Which can either get fixed, or deprecated.  Even if this 100% bug didn't exist, having this event firing for no reason seems a bit pointless...

**Comment by ahaldane** [comment]
Yes, that's a good idea.

In the meantime I also tried implementing the throttling. See https://github.com/ahaldane/matplotlib/commit/2dc661c50d5796479e2fb29008171da1dcc228ff. If anyone checks over it, try to confirm I am not creating race conditions.

I think we should combine your fix, plus throttling in some form, and we will be good.

**Comment by tacaswell** [comment]
This throttling seems like something that should be handled on the gtk side...

**Comment by WeatherGod** [comment]
@tacaswell, that's what I thought too. But my google-fu is weak...

**Comment by ahaldane** [comment]
I also looked a little for a GTK solution, but I didn't turn up anything.

By what I've read so far it doesn't seem possible: By my (weak!) understanding Glib creates various "event sources", including the idle event source, and each event source defines how often it is checked in its "prepare" function. Here's what it  [the docs](http://www.gtk-server.org/gtk1/glib/glib-the-main-event-loop.html#GSOURCEFUNCS) say about that:

> For idle sources, the prepare and check functions always return TRUE to indicate that the source is always ready to be processed. The prepare function also returns a timeout value of 0 to ensure that the poll() call doesn't block (since that would be time wasted which could have been spent running the idle function).

So as I understand glib, loops as fast as possible and skips any event sources that say they are not ready yet. However, idle sources are defined to always be ready.

**Comment by ahaldane** [comment]
Actually, thinking about it more, I think the idle callback is meant to be called as frequently as possible if it is defined - there is no way to have an idle callback which does not use 100% of the CPU. Glib looks through all event sources, and sleeps until the next event source says it will be ready, but idle sources are always ready so glib never sleeps when they're present.

So for matplotlib, I think that if the user has not specified an idle event, no idle event should be registered with glib, so glib can sleep. If the user supplies an idle callback, any throttling should be done by the user, and @OceanWolf 's solution is all that is needed. 

So in summary, I would implement  @OceanWolf 's solution, and also remove the "return True" that I added in the previous pull request. No extra throttling necessary.

**Comment by OceanWolf** [comment]
Hmm, thinking about it some more, if the user did want to use something like this, why not.

1. idle_event triggers callback(s)
2. idle_event calls a wait/sleep method for user specified amount of time
3. idle_event returns True and waits for next idle event...

**Comment by OceanWolf** [comment]
I don't know enough about GTK to know if that will work, i.e. if wait/sleep from the event_idle will block other events/processes.

Perhaps best to do the simple implementation to close this bug, and work on a better solution in a separate issue (along with @mdboom for your big plans :)?).

**Comment by tacaswell** [comment]
From what I understand I suspect that most things attached to idle in gtk
should be single shot triggers, if the client wants it to repeat they
should add a new event as part of the processing. Not sure how to did that
into the mpl version of events.

**Comment by efiring** [comment]
That's my understanding also.  We do not want to put in throttling, or a mechanism that keeps firing idle events continuously.  It looks like other toolkits are designed for this single-shot mode.  The only reason for having an mpl wrapper for the gui idle event would be to provide a uniform interface across backends; if we can't do this, we should rip out the idle event support entirely.  Then users needing it would have to use it in a manner that fits the chosen backend.
@OceanWolf, a sleep will block everything.

**Comment by ahaldane** [comment]
That makes sense.

It sounds like in any case ` idle_event` should be removed from the `mpl_connect` event handling code - that code is not set up to do one-shots since there is no mechanism to tell the backends to connect/disconnect their internal handlers. In addition, the wx and qt backends handle idle events differently. qt5 and tkagg backends don't handle idle events at all. wx backend uses the wx default, where idle events only get called once per idle period.

Now, maybe we still want to allow the user to add one-shot idle callbacks? If yes, I imagine we would need a new method ("on_idle"?), similar to `draw_idle` but for arbitrary callbacks.

Implementation for gtk might look something like this: https://github.com/ahaldane/matplotlib/commit/8e4ca70ae0c886c4de1fa95157f91323580bfca4, except that I couldn't quickly figure out how to keep track of multiple callbacks (for use in `destroy`) without introducing race conditions. `draw_idle` avoids race conditions by only allowing one callback at a time, which allows it to keep track of the callback id using atomic operations. (and I assume `FigureCanvasGTK.destroy` and the `idle_draw` callback run serially on the gtk thread otherwise I think there actually is a tiny race condition). Do we even need to remove these callbacks in destroy?

wx and qt seem better set up for one-shot callbacks. 

I'm not sure what the use case for one-shot idle would be. It wouldn't be animations, which seem to have been the original goal of idle_events (see `backend_bases.IdleEvent` docstring, plus example linked above).

**Comment by OceanWolf** [comment]
@ahaldane you could use a list of ids (create an empty list in __init__), add to the list in ``on_idle`` and remove in the callback, then do one big clear-up in destroy.

As for race-conditions, I think that might explain why one would want to return True... in python pseudo code, to avoid deadlock, for the GTK logic in python-pseudo-code:
```python
callback_failed = True
while callback_failed:
    callback_failed = callback()
```

**Comment by ahaldane** [comment]
The problem is that adding to the callback id list is non atomic. So the callback might run before the id is added to the list, and it would try to remove a nonexistent id.

I think we just have to use an explicit lock. Like this https://github.com/ahaldane/matplotlib/commit/eca8c4bb3d4ae299ac851391a3e9b787a84d6e90

I did a quick test, it seems to work. There is now a dependency on the threading module.

**Comment by OceanWolf** [comment]
@mdboom Any progress on this?  If we scrap this ``idle_event`` then this API change should go out with 2.0, ``color-overhaul`` (especially seeing as one of our examples depends on this).

**Comment by OceanWolf** [comment]
This either needs to get fixed, or have the fix that caused this backed out not cherrypicked for 1.5.

My position remains the same, that as we have here a documented feature with its own example on the live site, thus we should not just remove it without either a deprecation cycle, or doing it cleanly for 2.0.  In 1.4x GTK3 this feature didn't work, so no change, but for the rest AFAIK it worked.

**Comment by tacaswell** [comment]
@OceanWolf @ahaldane I have lost the thread of the discussion here.  Can one of you make a new issue about the outstanding issue?

**Comment by OceanWolf** [comment]
Okay working on creating a new issue about it now...
